<a href="https://colab.research.google.com/github/BelRom/HW_1_Analiz_data/blob/main/HW_1_Analiz_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

Видим что датасет использует после загрузки не оптимальный типы данных, откроем и используем оптимальные типы данных

In [ ]:
dtypes = {
    "row_id": "int64",
    "timestamp": "int64",
    "user_id": "int32",
    "content_id": "int16",
    "content_type_id": "boolean",
    "task_container_id": "int16",
    "user_answer": "int8",
    "answered_correctly": "int8",
    "prior_question_elapsed_time": "float32",
    "prior_question_had_explanation": "boolean"
}
train = pd.read_csv('train.csv', dtype=dtypes)

In [ ]:
print("Train size:", train.shape)
train.info()

Train size: (101230332, 10)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   row_id                          int64  
 1   timestamp                       int64  
 2   user_id                         int32  
 3   content_id                      int16  
 4   content_type_id                 boolean
 5   task_container_id               int16  
 6   user_answer                     int8   
 7   answered_correctly              int8   
 8   prior_question_elapsed_time     float32
 9   prior_question_had_explanation  boolean
dtypes: boolean(2), float32(1), int16(2), int32(1), int64(2), int8(2)
memory usage: 3.2 GB


В дата сете 101230332 строк и 10 столбцов, после приведения типов размер используемой памяти сократился в 2 раза

In [ ]:
train.describe()

row_id     timestamp       user_id    content_id  \
count  1.012303e+08  1.012303e+08  1.012303e+08  1.012303e+08   
mean   5.061517e+07  7.703644e+09  1.076732e+09  5.219605e+03   
std    2.922268e+07  1.159266e+10  6.197163e+08  3.866359e+03   
min    0.000000e+00  0.000000e+00  1.150000e+02  0.000000e+00   
25%    2.530758e+07  5.243436e+08  5.408116e+08  2.063000e+03   
50%    5.061517e+07  2.674234e+09  1.071781e+09  5.026000e+03   
75%    7.592275e+07  9.924551e+09  1.615742e+09  7.425000e+03   
max    1.012303e+08  8.742577e+10  2.147483e+09  3.273600e+04   

       task_container_id   user_answer  answered_correctly  \
count       1.012303e+08  1.012303e+08        1.012303e+08   
mean        9.040624e+02  1.376123e+00        6.251644e-01   
std         1.358302e+03  1.192896e+00        5.225307e-01   
min         0.000000e+00 -1.000000e+00       -1.000000e+00   
25%         1.040000e+02  0.000000e+00        0.000000e+00   
50%         3.820000e+02  1.000000e+00        1.000000e+00   
75%         1.094000e+03  3.000000e+00        1.000000e+00   
max         9.999000e+03  3.000000e+00        1.000000e+00   

       prior_question_elapsed_time  
count                 9.887879e+07  
mean                  2.542384e+04  
std                   2.122839e+04  
min                   0.000000e+00  
25%                   1.600000e+04  
50%                   2.100000e+04  
75%                   2.966600e+04  
max                   3.000000e+05

Проверим есть ли в нем нуливые значения

In [ ]:
train.isnull().any()

row_id                            False
timestamp                         False
user_id                           False
content_id                        False
content_type_id                   False
task_container_id                 False
user_answer                       False
answered_correctly                False
prior_question_elapsed_time        True
prior_question_had_explanation     True
dtype: bool

Нуливые значения есть только в 2 столбцах prior_question_elapsed_time и prior_question_had_explanation посмотрим количество нулевых значений

In [ ]:
train.prior_question_elapsed_time.isnull().sum()

2351538

In [ ]:
train.prior_question_had_explanation.isnull().sum()

392506

Посмотрим сколько уникальных пользователей

In [ ]:
print(f'Колличество пользователей {train.user_id.nunique()}')

Колличество пользователей 393656


Посмотрим какие значения и в каком количестве могут принемать столбцы content_type_id user_answer answered_correctly

In [ ]:
train.content_type_id.value_counts()

content_type_id
False    99271300
True      1959032
Name: count, dtype: Int64

In [ ]:
train.user_answer.value_counts()

user_answer
 0    28186489
 1    26990007
 3    26084784
 2    18010020
-1     1959032
Name: count, dtype: int64

In [ ]:
train.answered_correctly.value_counts()

answered_correctly
 1    65244627
 0    34026673
-1     1959032
Name: count, dtype: int64

In [ ]:
# Анализ доли правильных ответов для каждого пользователя
correct_ratio = train.groupby('user_id')['answered_correctly'].mean()
print("Средняя доля правильных ответов на пользователя:")
print(correct_ratio.head())

Средняя доля правильных ответов на пользователя:
user_id
115     0.695652
124     0.233333
2746    0.500000
5382    0.632812
8623    0.598214
Name: answered_correctly, dtype: float64


In [ ]:
user_stats = train.groupby('user_id')['prior_question_elapsed_time'].mean()
print("Среднее время ответа на пользователя:")
print(user_stats.head())

Среднее время ответа на пользователя:
user_id
115     19933.310547
124     18793.000000
2746    18055.554688
5382    36048.386719
8623    26107.408203
Name: prior_question_elapsed_time, dtype: float32


In [ ]:
task_analysis = train.groupby('task_container_id')['answered_correctly'].mean()
print("Средняя доля правильных ответов по task_container_id:")
print(task_analysis.head())

Средняя доля правильных ответов по task_container_id:
task_container_id
0    0.679709
1    0.526979
2    0.435514
3    0.543012
4    0.476876
Name: answered_correctly, dtype: float64


In [ ]:
# Вычисляем корреляцию
correlation = train.corr()

# Корреляция только с 'answered_correctly'
answered_correctly_corr = correlation['answered_correctly'].sort_values(ascending=False)

print("Корреляция признаков с answered_correctly:")
print(answered_correctly_corr)

Корреляция признаков с answered_correctly:
answered_correctly                1.000000
prior_question_had_explanation    0.261694
user_answer                       0.126471
task_container_id                 0.061129
timestamp                         0.025452
row_id                           -0.000303
user_id                          -0.000307
prior_question_elapsed_time      -0.007900
content_id                       -0.196960
content_type_id                  -0.436912
Name: answered_correctly, dtype: float64


Интерпретация результатов:
1. Высокая положительная корреляция:
prior_question_had_explanation (0.261694): Этот признак оказывает наибольшее положительное влияние на правильность ответа. Это означает, что если студент видел объяснение после предыдущего задания, вероятность того, что он правильно ответит на следующий вопрос, увеличивается.
2. Умеренная положительная корреляция:
user_answer (0.126471): Корреляция указывает, что выбранный пользователем ответ связан с правильностью ответа, но зависимость слабая.
task_container_id (0.061129): Небольшая положительная корреляция указывает, что группа заданий влияет на результаты. Возможно, определённые группы задач более понятны студентам.
3. Почти нулевая корреляция:
timestamp (0.025452): Время, прошедшее с начала сессии, практически не связано с правильностью ответа. Это может говорить о том, что успех на задании слабо зависит от длительности сессии.
4. Отрицательная корреляция:
prior_question_elapsed_time (-0.007900): Время, потраченное на предыдущий вопрос, слабо отрицательно связано с правильностью ответа. Это может говорить о том, что долгое обдумывание предшествующих вопросов не всегда приводит к успеху.
content_id (-0.196960): Негативная корреляция указывает, что некоторые типы контента могут быть более сложными для студентов.
content_type_id (-0.436912): Это самая сильная отрицательная корреляция. Признак content_type_id указывает смотрел лекцию студент или нет. Получается что если пользователь не смотрел лекцию результат будет хуже.

In [ ]:
lectures = pd.read_csv('lectures.csv')
display(lectures.info())
display(lectures.head())
lectures.shape


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   lecture_id  418 non-null    int64 
 1   tag         418 non-null    int64 
 2   part        418 non-null    int64 
 3   type_of     418 non-null    object
dtypes: int64(3), object(1)
memory usage: 13.2+ KB


None

lecture_id  tag  part           type_of
0          89  159     5           concept
1         100   70     1           concept
2         185   45     6           concept
3         192   79     5  solving question
4         317  156     5  solving question

(418, 4)

In [ ]:
lectures.isnull().any()

lecture_id    False
tag           False
part          False
type_of       False
dtype: bool

Нуливых значений в таблице нету

In [ ]:
display(lectures.type_of.value_counts())

type_of
concept             222
solving question    186
intention             7
starter               3
Name: count, dtype: int64

Посмотрев количество уникальных значений в столбце type_of приходим к выводу что
Тип starter встречается всего 3 раза. Это может быть вводная или начальная лекция/вопрос
concept: 222 Возможно, это лекции или вопросы, которые фокусируются на объяснении концепций.
solving question: 186 Это может быть связано с лекциями или вопросами, направленными на решение задач.
intention: 7  Возможно, это редкий тип, связанный с намерениями или целью обучения.
starter: 3 Это может быть вводная или начальная лекция/вопрос, предназначенная для старта обучения.

In [ ]:
print(lectures.lecture_id.nunique())
print(lectures.tag.nunique())

418
151


lecture_id совпадает с общим количеством строк, а уникальных тегов всего 151.

In [ ]:
questions = pd.read_csv('questions.csv')
display(questions.info())
display(questions.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   question_id     13523 non-null  int64 
 1   bundle_id       13523 non-null  int64 
 2   correct_answer  13523 non-null  int64 
 3   part            13523 non-null  int64 
 4   tags            13522 non-null  object
dtypes: int64(4), object(1)
memory usage: 528.4+ KB


None

question_id  bundle_id  correct_answer  part            tags
0            0          0               0     1   51 131 162 38
1            1          1               1     1       131 36 81
2            2          2               0     1  131 101 162 92
3            3          3               0     1  131 149 162 29
4            4          4               3     1    131 5 162 38

В датасете question_id и bundle_id совподают с номером строки

In [ ]:
questions.isnull().any()

question_id       False
bundle_id         False
correct_answer    False
part              False
tags               True
dtype: bool

In [ ]:
questions.isnull().sum()

question_id       0
bundle_id         0
correct_answer    0
part              0
tags              1
dtype: int64

In [ ]:
questions.dropna(subset=['tags'], inplace=True)

In [ ]:
display(questions.part.unique())
display(questions.correct_answer.unique())

array([1, 2, 3, 4, 5, 6, 7], dtype=int64)

array([0, 1, 3, 2], dtype=int64)

In [ ]:
# Группируем данные по 'part' и вычисляем среднее значение 'correct_answer'
correct_answer_by_part = questions.groupby('part')['correct_answer'].mean()
display(correct_answer_by_part)

part
1    1.502016
2    1.310868
3    1.499360
4    1.477415
5    1.454727
6    1.444261
7    1.547414
Name: correct_answer, dtype: float64

Как видим из таблицы больше всего правильных ответов в 7 часте и больше всего затруднений в 2